### PR TITLE
Iox #521 improve codecov

### LINKED ISSUE #521: Improve Pull-Request Process
## Brief feature description
In iceoryx are Pull-Requests often a long time open because of several inconsistencies:
- New contributors often don't know who is the committer with the best knowledge about the codebase where he is working on
- There are formal aspects expected from committer side which are not visible to the contributors (e.g. AAA Pattern in tests, which cases are needed to test? How to name tests properly?)
- The expectations regarding the scope of a PR are sometimes unclear (e.g. When is a PR too huge? What is expected on tests?)

Of course a Pull-Request is something iterative and the last thing we want is to have fast merges without reviewing it accordingly.
For more efficient Pull-Request reviews we need to improve the Contribution Process and give more information about it.

In the Contributing.md are already helpful hints listed: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
When i take a look at https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#testing i see not much info here, this can be a starting point to extend it.

Possible solutions (to be extended):
- [ ] extend the Contributor instructions with more hints
- [ ] add codeowners to iceoryx components which are automatically added to the Pull-Request. https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners
- [x] more automatic checks at Pull-Request like if all points on the PR checklist are marked
- [ ] add a PR Merge-Bot which ensures syncing with master and do automatic checks
- [x] add labels to PR template so that Contributors can set them
- [x] enable annotations for codecov.io

Maybe we can have here a starting point with a checklist (unordered and incomplete):
- [ ] build everything and execute tests: `./tools/iceoryx_build_test.sh build-strict build-all out-of-tree test clean`
- [ ] are unittests complete? [Link](https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#testing)
- [ ] documentation in doxygen complete? [Link](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/aspice_swe3_4/swe_docu_guidelines.md)
- [ ] clang-format the code?
- [ ] naming correct? See: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#naming-conventions
- [ ] fileheader correct? (company)
- [ ] use `const &` in function arguments wherever possible
- [ ] avoid using fundamental types like `int`, use fixed width integer types like `uint32_t`
- [ ] always initialize member variables and use uppercase literals if possible `uint32_t foo{0};`
- [ ] no commented code
- [ ] Are the labels in the PR properly set? (E.g. test, refactoring etc.)
- [x] ...

checklist for tests in Contributing.md(incomplete):
- [x] use AAA pattern for testcase structure https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80
- [x] use `sut` as member name for the object to test
- [x] prefer to use stack-allocated objects for testing, if only heap is possible then wrap them in smart-pointers to ensure correct cleanup and avoid memleaks
- [x] create a meaningful name for your testcase, e.g. `TEST_F(MemoryManager_test, PerformSomethingAndExpectThis)`
- [x] don't test multiple aspects in one testcase, prefer to split up in different little testcases
- [x] avoid usage of `new` in tests as much as possible because you need to `delete` it afterwards
- [x] test negative cases too
- [x] possible abstract testcases: zero, one, many, corner cases, limits
- [x] ...

Feel free to make proposals and extend the checklist.
@eclipse-iceoryx/technology-iceoryx-committers

### PR DESCRIPTION
**Pull request by dkroenke**
## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer
- enabling comments on PRs with changing code coverage
- set thresholds to prevent merges with too much uncovered lines
- add option to build tests using `TEST_WITH_ADDITIONAL_USER`
- integrate tests with additional users in Github CI

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [x] All checkboxes in the PR checklist are checked or crossed out
1. [x] Merge

## References

- Partly Closes #521

### CODE REVIEW

File: iceoryx_utils/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
-# Copyright (c) 2019, 2020 by Robert Bosch GmbH, Apex.AI Inc. All rights reserved.
+# Copyright (c) 2019 - 2021 by Robert Bosch GmbH. All rights reserved.
```
Comment (dkroenke) [inline iceoryx_utils/CMakeLists.txt:1]: @marthtz would you check if the dates are correct here? I set them based on the git file history.
Same for the files below.

File: tools/iceoryx_build_test.sh [hunk 2]
```diff
@@ -185,8 +193,9 @@ while (( "$#" )); do
         echo "    build-all             Build all extensions and all examples"
         echo "    out-of-tree           Out-of-tree build for CI build"
         echo "    build-strict          Build is performed with '-Werror'"
-        echo "    build-shared         Build shared libs (iceoryx is build as static lib per default)"
+        echo "    build-shared          Build shared libs (iceoryx is build as static lib per default)"
         echo "    build-test            Builds all tests (doesn't run)"
+        echo "    test-add-user         Create additional useraccounts in system for testing access controll (default off)"
```
Comment (marthtz) [inline tools/iceoryx_build_test.sh:198]: Some typos: `Create additional user accounts in system for testing access control (default: off)"`

File: tools/iceoryx_build_test.sh [hunk 3]
```diff
@@ -175,7 +182,8 @@ while (( "$#" )); do
         echo "    iceoryx_build_test.sh [--build-dir <dir>] [<args>]"
         echo "Options:"
         echo "    -b --build-dir         Specify a non-default build directory"
-        echo "    -c --coverage         Builds gcov and generate a html/xml report. Possible arguments: 'all', 'unit', 'integration'"
+        echo "    -c --coverage         Builds gcov and generate a html/xml report.
```
Comment (marthtz) [inline tools/iceoryx_build_test.sh:185]: Isn't a quotation mark `"` missing at EOL?

File: tools/iceoryx_build_test.sh [hunk 4]
```diff
@@ -175,7 +182,8 @@ while (( "$#" )); do
         echo "    iceoryx_build_test.sh [--build-dir <dir>] [<args>]"
         echo "Options:"
         echo "    -b --build-dir         Specify a non-default build directory"
```
Comment (marthtz) [inline tools/iceoryx_build_test.sh:184]: While we're here... looks like an extra space before `Specify`?

File: tools/iceoryx_build_test.sh [hunk 5]
```diff
@@ -185,8 +193,9 @@ while (( "$#" )); do
         echo "    build-all             Build all extensions and all examples"
         echo "    out-of-tree           Out-of-tree build for CI build"
         echo "    build-strict          Build is performed with '-Werror'"
-        echo "    build-shared         Build shared libs (iceoryx is build as static lib per default)"
+        echo "    build-shared          Build shared libs (iceoryx is build as static lib per default)"
```
Comment (marthtz) [inline tools/iceoryx_build_test.sh:196]: Should be `iceoryx is built`.

File: tools/gcov/lcov_generate.sh [hunk 6]
```diff
@@ -25,16 +27,15 @@ case "$2" in
     "initial")
         lcov -c -i -d $BUILD_FOLDER -o $OUTPUT_FOLDER/iceoryx_init.info --no-external --rc lcov_branch_coverage=1
         ;;
-    "capture")
+    "scan")
```
Comment (marthtz) [inline tools/gcov/lcov_generate.sh:30]: It now says `scan` here but it's still `capture` in line 20, is that correct?

File: iceoryx_utils/CMakeLists.txt (hunk 1 above)
Comment (marthtz) [reply iceoryx_utils/CMakeLists.txt:1]: Yep, all good!

File: .github/workflows/build-test.yml [hunk 7]
```diff
@@ -36,25 +36,25 @@ jobs:
...
-            $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh build-strict build-all out-of-tree build-shared
+            sudo $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh build-strict build-all out-of-tree build-shared test-add-user
 
        - name: Build debian package
          run: |
-            $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh package
+            sudo $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh package
 
-       - name: Run timeless tests
+       - name: Run all tests, excluding timing_tests
```
Comment (marthtz) [inline .github/workflows/build-test.yml:45]: So, timing tests are run separately now and not in the default test set?

File: .github/workflows/build-test.yml [hunk 8]
```diff
@@ -36,25 +36,25 @@ jobs:
...
          run: |
            cd $GITHUB_WORKSPACE/build
-           make all_tests
+           sudo make all_tests
 
        - name: Run timing tests
          run: |
            cd $GITHUB_WORKSPACE/build
-           make timing_tests
+           sudo make timing_tests
```
Comment (marthtz) [inline .github/workflows/build-test.yml:53]: Will this also include the timing tests in the coverage?

File: .github/workflows/coverage_doc.yml [hunk 9]
```diff
@@ -31,31 +31,27 @@ jobs:
...
+        with:
+          fetch-depth: 0
 
       - name: Build, test and generate gcov report
         run: |
+          mkdir $GITHUB_WORKSPACE/lcov_results
+          mkdir $GITHUB_WORKSPACE/lcov_results/unittest
           # add flag to include coverage
-          $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh build-strict debug build-all out-of-tree -c unit
+          sudo $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh build-strict build-all test-add-user -c unit
```
Comment (elfenpiff) [inline .github/workflows/coverage_doc.yml:42]: @mossmaurice mentioned in another context that timing_tests are not considered when calculating the test coverage. Is this correct or do we have to add them maybe here?

File: .github/workflows/sanitize.yml [hunk 10]
```diff
@@ -30,12 +30,13 @@ jobs:
       - name: Build
         run: |
           # build tests , dont run them
-          $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh build-strict build-all sanitize clang
+          sudo $GITHUB_WORKSPACE/tools/iceoryx_build_test.sh build-strict build-all sanitize clang test-add-user
```
Comment (elfenpiff) [inline .github/workflows/sanitize.yml:33]: When we build the sanity check do we also build all examples to ensure that they do not have a compile problem in sanitize mode?

File: tools/iceoryx_build_test.sh [hunk 11]
```diff
@@ -95,6 +97,11 @@ while (( "$#" )); do
         TEST_FLAG="ON"
         shift 1
         ;;
+    "test-add-user")
+        TEST_FLAG="ON"
```
Comment (elfenpiff) [inline tools/iceoryx_build_test.sh:101]: I wouldn't enable the test flag here. If you would like to have a test user but no tests it is totally fine. I am just afraid that we go back to our old cmake state where nearly every flag had side effects and some of them enabled stuff and other disabled them.

File: .github/workflows/build-test.yml (hunk 7 above)
Comment (dkroenke) [reply .github/workflows/build-test.yml:45]: Yes, the idea is that we collect all tests with timing constraints into one target to execute them separately in an environment with defined hardware specs. Currently it seems that the timing_tests does not produce too much trouble on while running on GitHub CI.

File: .github/workflows/build-test.yml (hunk 8 above)
Comment (dkroenke) [reply .github/workflows/build-test.yml:53]: Currently not because the `TIMING_TEST` macro is used in module -and integrationtests which would distort the coverage value for unittests if we add this here now.
Possible solution could be to separate the Timing_tests into different targets for module -and integrationtests.

File: .github/workflows/coverage_doc.yml (hunk 9 above)
Comment (JimmyBch) [reply .github/workflows/coverage_doc.yml:42]: Hello, if we want to activate timing test in coverage, I think we should remove the filter in tests.cmake located in folder iceoryx_meta at line 34 for module tests and line 39 for integration tests. 

I tried to add also in script run_all_tests.sh at line 135 another line with make timing_tests and it was working as well :)

File: .github/workflows/build-test.yml (hunk 8 above)
Comment (dkroenke) [reply .github/workflows/build-test.yml]: Ok i split up the timing_test targets and include them in the coverage report. We need still to have the separation of timing_tests otherwise we will have trouble on MacOS CI build.

Comment (JimmyBch) [reply .github/workflows/build-test.yml]: Should we have to use a different cmd to build and run the test locally on ubuntu or we can keep the current one ?

./tools/iceoryx_build_test.sh clean build-strict debug test

Comment (dkroenke) [reply .github/workflows/build-test.yml]: You can keep the current one, this executes the tests with scope `all`

File: .github/workflows/sanitize.yml (hunk 10 above)
Comment (dkroenke) [reply .github/workflows/sanitize.yml]: all examples are build with the sanitizer flags

**Comment by JimmyBch** [comment]
@dkroenke I noticed that utils folder is not taken in consideration when I check locally the coverage. Will this PR fix the issue as well ?

**Comment by dkroenke** [comment]
> @dkroenke I noticed that utils folder is not taken in consideration when I check locally the coverage. Will this PR fix the issue as well ?

Can you post the command you are using locally? I tested on Ubuntu 18 and 20 with different scopes (all and unit) and can still see the coverage for iceoryx_utils in the generated html report.

**Comment by JimmyBch** [comment]
> > @dkroenke I noticed that utils folder is not taken in consideration when I check locally the coverage. Will this PR fix the issue as well ?
> 
> Can you post the command you are using locally? I tested on Ubuntu 18 and 20 with different scopes (all and unit) and can still see the coverage for iceoryx_utils in the generated html report.

Here is the cmd ./tools/iceoryx_build_test.sh clean build-strict debug -c

I am looking after that into build/lcov/ but I have only posh files available.

**Comment by dkroenke** [comment]
> Here is the cmd ./tools/iceoryx_build_test.sh clean build-strict debug -c
> 
> I am looking after that into build/lcov/ but I have only posh files available.

Is the generated html also not showing the coverage for utils? Can you see in the log output that the utils are processed? I tried it with your command and can still see the utils coverage.
It is important that the test run is successful because that generates the gcda files with the coverage information for each source file.

File: .github/workflows/build-test.yml [hunk 12]
```diff
@@ -36,21 +35,17 @@ jobs:
...
+       - name: Run all tests
          run: |
            cd $GITHUB_WORKSPACE/build
-           make all_tests
-
-       - name: Run timing tests
-         run: |
-           cd $GITHUB_WORKSPACE/build
-           make timing_tests
+           tools/run_all_tests.sh all
```
Comment (elBoberido) [inline .github/workflows/build-test.yml:48]: since the `run_all_tests.sh` script now takes a parameter to run all tests, shouldn't this be renamed to `run_tests.sh`?

File: tools/add_test_users.sh [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+# Use at your own risk!
+
+set -e
+
+#====================================================================================================
+#==== Step : Create local test users and groups for testing access control  =========================
+#====================================================================================================
+
+
+if [ $(getent group roudi_test1) ] && [ $(getent group roudi_test2) ] && [ $(getent group roudi_test3) ]; then
```
Comment (elBoberido) [inline tools/add_test_users.sh:29]: Just adding all users if just one is missing, might also to problems. It must be checked if the specific user is already existing before adding it. Below, there is a `USERS` array which could also be used to check if the user is already existing and if not create one. The sudo check could be done right before the user is added.

File: tools/iceoryx_build_test.sh [hunk 14]
```diff
@@ -95,6 +97,11 @@ while (( "$#" )); do
         TEST_FLAG="ON"
         shift 1
         ;;
+    "test-add-user")
+        TEST_ADD_USER="ON"
+        echo "Warning: Tests are running with user accounts 'roudi_testX', please make sure that add_test_users.sh has run before."
+        shift 1
+        ;;
```
Comment (elBoberido) [inline tools/iceoryx_build_test.sh:104]: This can easily be overlooked because there is a wall of output from the tests when the script is run. Could you maybe call the `add_test_users.sh` with a `check` parameter which aborts the script if not all users are available?

File: tools/add_test_users.sh [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    echo "users and groups for testing already exist"
+else
+    if [ "$EUID" -ne 0 ]; then
+        echo "Please run the script as root with sudo"
+        exit 1
+    fi
+    USERS="roudi_test1 roudi_test2 roudi_test3"
```
Comment (elBoberido) [inline tools/add_test_users.sh:36]: oh, can we have a `iox` prefix?

File: tools/run_all_tests.sh [hunk 16]
```diff
@@ -132,16 +122,24 @@ execute_test() {
 
     case $test_scope in
     "all")
+        make all_tests
+        make integration_tests
+        make timing_integrationtests
```
Comment (elBoberido) [inline tools/run_all_tests.sh:127]: does `all_tests` only run all module_tests?

File: tools/run_all_tests.sh [hunk 17]
```diff
@@ -132,16 +122,24 @@ execute_test() {
...
+        ;;
+    "no_timing_test")
         make all_tests
         ;;
     "unit")
         make module_tests
+        make timing_moduletests
```
Comment (elBoberido) [inline tools/run_all_tests.sh:134]: this is a bit inconsistent. `mudule_tests` is separated with an underscore but not in `timing_moduletests`

File: tools/run_all_tests.sh (hunk 16 above)
Comment (dkroenke) [reply tools/run_all_tests.sh]: No this target runs all module -and integrationtests without timing_tests

Comment (elBoberido) [reply tools/run_all_tests.sh]: isn't the `make integration_tests` superfluous? Or should this be `make timing_module_tests`?

**Review by elBoberido — COMMENTED** [review]
Looks good. Just one final clarification then I could approve this.

**Review by elBoberido — APPROVED** [review]

File: iceoryx_posh/test/moduletests/test_mepoo_segment_management.cpp [hunk 18]
```diff
@@ -69,8 +69,8 @@ class SegmentManager_test : public Test
     SegmentConfig getSegmentConfig()
     {
         SegmentConfig config;
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_mepoo_segment_management.cpp:71]: Please adjust copyright header.

**Review by elfenpiff — CHANGES_REQUESTED** [review]

**Review by elfenpiff — APPROVED** [review]
